python/ray/util/sgd/torch/examples/segmentation/train_segmentation.py

```diff
@@ -0,0 +1,266 @@
+
+    if len(losses) == 1:
+        return losses["out"]
+
+    return losses["out"] + 0.5 * losses["aux"]
+
+
+class SegOperator(TrainingOperator):
+    def setup(self, config):
+        pass
+        # Currently, this LR scheduler is broken due to
+        # https://github.com/pytorch/pytorch/issues/32639. But theoretically,
+        # other LR schedulers should work.
+        # datalen = len(self.train_dataloader)
+        # args = config["args"]
+        # self.scheduler = torch.optim.lr_scheduler.LambdaLR(
+        #     self.optimizer,
+        #     lambda x: (1 - x / (datalen * args.epochs))**0.9
+        # )
+
+    def train_batch(self, batch, batch_info):
+        image, target = batch
+        image, target = image.to(self.device), target.to(self.device)
+        output = self.model(image)
+        loss = criterion(output, target)
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+        # self.scheduler.step()
+        lr = self.optimizer.param_groups[0]["lr"]
+        return {"loss": loss.item(), "lr": lr, "num_samples": len(batch)}
+
+    def validate(self, data_loader, info=None):
+        self.model.eval()
+        confmat = utils.ConfusionMatrix(self.config["num_classes"])
+        with torch.no_grad():
+            for image, target in data_loader:
+                image, target = image.to(self.device), target.to(self.device)
+                output = self.model(image)
+                output = output["out"]
+
+                confmat.update(target.flatten(), output.argmax(1).flatten())
+
+            confmat.reduce_from_all_processes()
+
+        return confmat
+
+
+def optimizer_creator(model, config):
+    args = config["args"]
+    model_without_ddp = model.module if hasattr(model, "module") else model
+    params_to_optimize = [
+        {
+            "params": [
+                p for p in model_without_ddp.backbone.parameters()
+                if p.requires_grad
+            ]
+        },
+        {
+            "params": [
+                p for p in model_without_ddp.classifier.parameters()
+                if p.requires_grad
+            ]
+        },
+    ]
+    if args.aux_loss:
+        params = [
+            p for p in model_without_ddp.aux_classifier.parameters()
+            if p.requires_grad
+        ]
+        params_to_optimize.append({"params": params, "lr": args.lr * 10})
+    return torch.optim.SGD(
+        params_to_optimize,
+        lr=args.lr,
+        momentum=args.momentum,
+        weight_decay=args.weight_decay)
+
+
+def model_creator(config):
+    args = config["args"]
+    model = torchvision.models.segmentation.__dict__[args.model](
+        num_classes=config["num_classes"],
+        aux_loss=args.aux_loss,
+        pretrained=args.pretrained)
+    # We usually don't need to do this, but we're enabling batch norm
+    # so we want to convert everything to GPU tensors first.
```

Comment:
So ideally we should be able to do this after the model gets returned and put onto the GPU if GPU training is enabled? Ig we should think through the API here better.

Reply:
actually, maybe it doesn't matter - I'm going to try this.